tools/get_coverage_files.bzl

```diff
@@ -0,0 +1,82 @@
+load("//cc:defs.bzl", "LIBRARY", "TEST_LIBRARY")
+load("@bazel_skylib//rules:common_settings.bzl", "BuildSettingInfo")
+
+FilesInfo = provider(
+    fields = {
+        "files": "files of the target",
+    },
+)
+
+def _get_cov_files_impl(target, ctx):
+    files = []
+
+    if not CcInfo in target:
+        return [FilesInfo(files = [])]
+
+    tags = getattr(ctx.rule.attr, "tags", [])
+    if not LIBRARY in tags and not TEST_LIBRARY in tags:
+        return [FilesInfo(files = [])]
+
+    if hasattr(ctx.rule.attr, "srcs"):
+        for src in ctx.rule.attr.srcs:
+            for file in src.files.to_list():
+                if file.is_source and not file.path.startswith(ctx.genfiles_dir.path):
+                    files.append(file)
+
+    if hasattr(ctx.rule.attr, "hdrs"):
+        for hdr in ctx.rule.attr.hdrs:
+            for file in hdr.files.to_list():
+                if not file.path.startswith(ctx.genfiles_dir.path):
+                    files.append(file)
+
+    return [FilesInfo(files = files)]
+
+_get_cov_files = aspect(
+    implementation = _get_cov_files_impl,
+)
+
+def _gen_sonar_cfg_impl(ctx):
+    all_files_bash = ""
+    all_files = []
+
+    for target in ctx.attr.targets:
+        for file in target[FilesInfo].files + ctx.files.test_srcs:
+            if file.path not in all_files:
+                all_files_bash += file.path
+                all_files_bash += " "
+                all_files.append(file.path)
+
+    out = ctx.actions.declare_file("sonar-project.properties")
+    ctx.actions.run_shell(
+        outputs = [out],
+        command = """
+        echo 'sonar.sourceEncoding=UTF-8' >> {out_file}
+        echo 'sonar.sources=\\' >> {out_file}
+        first=1
+        for file in {all_files}; do
+            if [[ $first -eq 1 ]]
+            then
+                first=0
+            else
+                echo ',\\' >> {out_file}
+            fi
+            echo -n "  {root_dir}/$file" >> {out_file}
```

Comment:
So is what we're doing with the `root_dir` var making sure that the sources original absolute path?

Reply:
Yes, that's right.

Reply:
I think that might cause a problem for caching since the `sonar-project.proprties` will be different from machine to machine.

But we can probably live with that for now.